### PR TITLE
Rewrite docs for custom predicates

### LINKED ISSUE #1161: Rewrite docs for "Custom Route Predicates"
The section "Custom Route Predicates" in narr/urldispatch.rst still uses the (as of 1.5) deprecated method of adding route predicates via `add_route/add_view(custom_predicates=...)`. It should be rewritten to use the new `Configurator.add_view_predicate`/`Configurator.add_route_predicate`.

### PR DESCRIPTION
**Pull request by merwok**
Fixes #1161

### CODE REVIEW

File: docs/narr/urldispatch.rst [hunk 1]
```diff
@@ -1100,52 +1097,73 @@ was matched (see :class:`pyramid.interfaces.IRoute` for the API of such a route
...
+            self.allowed = tuple(val[0:])
+
+        def text(self):
+            args = (self.segment_name,) + self.allowed
+            return 'any_of = %s' % (args,)
+
+        phash = text
+
+        def __call__(self, info, request):
+            return info['match'][self.segment_name] in self.allowed
```
Comment (merwok) [inline docs/narr/urldispatch.rst]: I haven’t tested this code yet

File: docs/narr/urldispatch.rst [hunk 2]
```diff
@@ -1100,52 +1097,73 @@ was matched (see :class:`pyramid.interfaces.IRoute` for the API of such a route
...
+An instance of the class defined above is a predicate which ensures that the
 match value named ``segment_name`` is in the set of allowable values
-represented by ``allowed``.  We use this ``any_of`` function to generate a
-predicate function named ``num_one_two_or_three``, which ensures that the
-``num`` segment is one of the values ``one``, ``two``, or ``three`` , and use
-the result as a custom predicate by feeding it inside a tuple to the
-``custom_predicates`` argument to
-:meth:`~pyramid.config.Configurator.add_route`.
+represented by ``allowed``.  We register this class as a *predicate
+factory* with the ``any_of`` argument name, then we can use that new
```
Comment (merwok) [inline docs/narr/urldispatch.rst]: I could remove the explanation about the factory here, as the other docs (narr/hooks.rst) goes into more detail.

File: docs/narr/urldispatch.rst [hunk 3]
```diff
@@ -1100,52 +1097,73 @@ was matched (see :class:`pyramid.interfaces.IRoute` for the API of such a route
...
+keyword argument with :meth:`~pyramid.config.Configurator.add_route` to
+have Pyramid instantiate the class (the ``('num', 'one', 'two',
+'three')`` value is passed to the constructor as the argument ``val``)
+and use the resulting instance as a custom predicate to check if
+incoming requests match the route or not.
```
Comment (merwok) [inline docs/narr/urldispatch.rst]: This sentence is a bit long, please advise on better wording!

File: docs/narr/urldispatch.rst [hunk 4]
```diff
@@ -1199,61 +1225,42 @@ route in a set of route predicates:
...
-    classmethod_predicate.__text__ = 'my classmethod predicate'
-    classmethod_predicate = classmethod(classmethod_predicate)
+The ``phash`` method should return a string that uniquely identifies a
+specific predicate (the name means *predicate hash*).  A good way to do
+that is to use the same argument name and value that in the call to
+``add_route``, like in the examples above.
 
-The same will work with ``staticmethod``, using ``staticmethod`` instead of
-``classmethod``.
+.. XXX add note about predicate matching for boolean custom predicates?
```
Comment (merwok) [inline docs/narr/urldispatch.rst]:
```suggestion

```

File: docs/narr/viewconfig.rst [hunk 5]
```diff
@@ -506,20 +508,22 @@ configured view.
...
 
-  If ``custom_predicates`` is not specified, no custom predicates are used.
+  .. deprecated:: 1.5
+     See section below for new-style custom predicates.
 
-``predicates``
-  Pass a key/value pair here to use a third-party predicate registered via
-  :meth:`pyramid.config.Configurator.add_view_predicate`.  More than one
-  key/value pair can be used at the same time.  See
+``**predicates``
```
Comment (merwok) [inline docs/narr/viewconfig.rst]: IMO adding the `**` makes it clearer that we’re talking about varkwargs here, but if that style is never used in the doc I could revert the change.

File: src/pyramid/config/routes.py [hunk 6]
```diff
@@ -283,6 +283,8 @@ def add_route(
 
           .. versionadded:: 1.4a4
 
+          .. deprecated:: 2.0
```
Comment (merwok) [inline src/pyramid/config/routes.py]: The docs build, but I haven’t checked the output.  I will compare with the method below, where the whole doc block is inside the deprecated directive.

File: docs/narr/viewconfig.rst [hunk 7]
```diff
@@ -506,20 +508,22 @@ configured view.
...
   resource and/or the request.  If all callables return ``True``, the
   associated view callable will be considered viable for a given request.
+  This parameter is kept around for backward compatibility.
 
-  If ``custom_predicates`` is not specified, no custom predicates are used.
+  .. deprecated:: 1.5
+     See section below for new-style custom predicates.
 
-``predicates``
-  Pass a key/value pair here to use a third-party predicate registered via
```
Comment (merwok) [inline docs/narr/viewconfig.rst]: I didn’t like `third-party` here, as custom predicates can be written by the dev, who is the first party!

File: docs/narr/urldispatch.rst (hunk 2 above)
Comment (stevepiercy) [reply docs/narr/urldispatch.rst]: I think it is appropriate here. However, I would break this into shorter sentences as it is quite long.

File: src/pyramid/config/routes.py (hunk 6 above)
Comment (stevepiercy) [reply src/pyramid/config/routes.py]: Tested. Output looks fine.

File: src/pyramid/config/routes.py [hunk 8]
```diff
@@ -302,14 +304,14 @@ def add_route(
               :ref:`custom_route_predicates` for more information about
               ``info``.
 
-        predicates
+        \*\*predicates
 
-          Pass a key/value pair here to use a third-party predicate
+          Pass extra keyword parameters to use custom or third-party predicates
```
Comment (stevepiercy) [inline src/pyramid/config/routes.py]:
```suggestion
          Pass extra keyword parameters to use custom predicates
```

File: src/pyramid/config/views.py [hunk 9]
```diff
@@ -740,14 +742,14 @@ def wrapper(context, request):
                 obsoletes this argument, but it is kept around for backwards
                 compatibility.
 
-        view_options
+        \*\*view_options
 
-          Pass a key/value pair here to use a third-party predicate or set a
-          value for a view deriver. See
+          Pass extra keyword parameters to use custom or third-party predicates
```
Comment (stevepiercy) [inline src/pyramid/config/views.py]:
```suggestion
          Pass extra keyword parameters to use custom predicates
```

File: docs/narr/urldispatch.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (merwok) [reply docs/narr/urldispatch.rst]: If I remember right, I had an issue configuring some debug views in my project with a custom predicate `is_authenticated=True/False` (because the `persmission` parameter to `view_config` is not a predicate param), and could not get it to work at some point until I deleted something, making one view config less specific than the other and so matching what I wanted.  It’s not related to *custom* predicates specifically though, so should just remove the line here.

File: docs/narr/viewconfig.rst [hunk 10]
```diff
@@ -506,20 +508,22 @@ configured view.
...
   predefined predicates as necessary.  Each custom predicate callable should
   accept two arguments, ``context`` and ``request``, and should return either
   ``True`` or ``False`` after doing arbitrary evaluation of the context
   resource and/or the request.  If all callables return ``True``, the
   associated view callable will be considered viable for a given request.
+  This parameter is kept around for backward compatibility.
 
-  If ``custom_predicates`` is not specified, no custom predicates are used.
+  .. deprecated:: 1.5
+     See section below for new-style custom predicates.
```
[the remark this answers is not in the document]
Comment (merwok) [reply docs/narr/viewconfig.rst]: :crying_cat_face:

File: docs/quick_tour.rst [hunk 11]
```diff
@@ -478,7 +478,7 @@ more to offer:
 
 - One route leading to multiple views, based on information in the request or
   data such as ``request_param``, ``request_method``, ``accept``, ``header``,
-  ``xhr``, ``containment``, and ``custom_predicates``
+  ``xhr``, ``containment``, and custom predicates.
```
[the remark this answers is not in the document]
Comment (merwok) [reply docs/quick_tour.rst]: New text references the concept of custom predicates, old text was referring to the specific `custom_predicates` parameter.

File: docs/narr/urldispatch.rst (hunk 3 above)
Comment (merwok) [reply docs/narr/urldispatch.rst]: Ah, I am not a fan of long lines (harder editor navigation, less nice to read, sometimes makes worse diffs), but I will try to follow the existing convention.  Thanks for the text, I will merge our versions.

File: docs/narr/viewconfig.rst (hunk 7 above)
Comment (merwok) [reply docs/narr/viewconfig.rst]: OK to change to `custom predicate` in all the spots you noted!

I still don’t get why third-party packages are mentioned.  The app I am writing to test Pyramid 2 has no third-party custom predicates, only mine.

File: docs/narr/urldispatch.rst [hunk 12]
```diff
@@ -1100,52 +1097,73 @@ was matched (see :class:`pyramid.interfaces.IRoute` for the API of such a route
...
 
 .. code-block:: python
     :linenos:
 
-    def any_of(segment_name, *allowed):
-        def predicate(info, request):
-            if info['match'][segment_name] in allowed:
-                return True
-        return predicate
+    class AnyOfPredicate:
```
Comment (merwok) [inline docs/narr/urldispatch.rst]: Note that these example classes do not redundantly inherit from `object`, even though I saw that many parts of the current docs still have that.

Comment (mmerickel) [reply docs/narr/urldispatch.rst]: It's fine to drop the object parent, I'd accept a PR that did it consistently to the whole codebase but it's unimportant right now - new stuff is free to not use it like in this PR.

**Comment by merwok** [comment]
Please note that I have made multiple commits for ease of review (and possibly revert), but the result should be a simple squashed changeset.

File: HISTORY.rst [hunk 13]
```diff
@@ -2530,7 +2530,7 @@ Features
      @view_config(abc=1)
 
   Similar features exist for ``add_route``, and ``add_subscriber``.  See
-  "Adding A Third Party View, Route, or Subscriber Predicate" in the Hooks
+  "Adding A Custom View, Route, or Subscriber Predicate" in the Hooks
   chapter for more information.
```
Comment (merwok) [inline HISTORY.rst]: I have avoided changing the other historical changelogs, but I changed here to avoid a dead reference.

File: src/pyramid/config/routes.py [hunk 14]
```diff
@@ -40,12 +40,9 @@ def add_route(
         inherit_slash=None,
         **predicates
     ):
-        """ Add a :term:`route configuration` to the current
-        configuration state, as well as possibly a :term:`view
-        configuration` to be used to specify a :term:`view callable`
-        that will be invoked when this route matches.  The arguments
-        to this method are divided into *predicate*, *non-predicate*,
-        and *view-related* types.  :term:`Route predicate` arguments
```
Comment (merwok) [inline src/pyramid/config/routes.py]: Unrelated to the current PR, but thought it was easier to cleanup that to open a separate ticket.  Easy to revert if you request so!

File: src/pyramid/config/views.py [hunk 15]
```diff
@@ -769,7 +768,7 @@ def wrapper(context, request):
                     'Configurator.add_view is deprecated as of Pyramid 1.5. '
                     'Use "config.add_view_predicate" and use the registered '
                     'view predicate as a predicate argument to add_view '
-                    'instead. See "Adding A Third Party View, Route, or '
+                    'instead. See "Adding A Custom View, Route, or '
```
Comment (stevepiercy) [inline src/pyramid/config/views.py]: Please use a `:ref:` instead of literal text.

```
:ref:`registering_custom_predicates`
```

File: src/pyramid/config/routes.py [hunk 16]
```diff
@@ -321,7 +319,7 @@ def add_route(
                     'Configurator.add_route is deprecated as of Pyramid 1.5. '
                     'Use "config.add_route_predicate" and use the registered '
                     'route predicate as a predicate argument to add_route '
-                    'instead. See "Adding A Third Party View, Route, or '
+                    'instead. See "Adding A Custom View, Route, or '
                     'Subscriber Predicate" in the "Hooks" chapter of the '
                     'documentation for more information.'
```
[the remark this answers is not in the document]
Comment (merwok) [reply src/pyramid/config/routes.py]: This is a warning message, not something extracted to docs.

[the remark this answers is not in the document]
Comment (stevepiercy) [reply src/pyramid/config/routes.py]: Oops. Sorry about that.

File: docs/narr/urldispatch.rst [hunk 17]
```diff
@@ -1121,22 +1121,20 @@ See this code:
...
-'three')`` value is passed to the constructor as the argument ``val``)
-and use the resulting instance as a custom predicate to check if
-incoming requests match the route or not.
+We register this class as a :term:`predicate factory` with the ``any_of`` keyword argument name.
+Then we use that new keyword argument with :meth:`~pyramid.config.Configurator.add_route`.
+When the route is requested, Pyramid instantiates the ``AnyOfPredicate`` class using the value passed to the ``any_of``
+argument.  The resulting instance is a :term:`predicate`.  It will determine whether incoming requests satisfy its condition.
+In the example above, a request for ``/three`` would match the route's URL pattern and satisfy the route's predicate
+because ``three`` is one of the allowed values, so the route would be matched.  However a request for ``/millions`` will
+match the route's URL pattern but would not satisfy the route's predicate, and the route would not be matched.
```
[the remark this answers is not in the document]
Comment (stevepiercy) [reply docs/narr/urldispatch.rst]: We previously had line breaks at 79 columns for narrative docs. It has been changed to one sentence per line. These examples are 110+ characters which breaks both the old and new guidelines.

File: src/pyramid/config/routes.py [hunk 18]
```diff
@@ -40,12 +40,9 @@ def add_route(
         inherit_slash=None,
         **predicates
     ):
-        """ Add a :term:`route configuration` to the current
-        configuration state, as well as possibly a :term:`view
-        configuration` to be used to specify a :term:`view callable`
-        that will be invoked when this route matches.  The arguments
-        to this method are divided into *predicate*, *non-predicate*,
-        and *view-related* types.  :term:`Route predicate` arguments
+        r""" Add a :term:`route configuration` to the current configuration state.
```
Comment (mmerickel) [inline src/pyramid/config/routes.py]: I do not like this pattern of using an r-string so that you can make `**predicates` work. Is there some other way to escape those two characters like `\\*\\*` ?

File: docs/narr/urldispatch.rst [hunk 19]
```diff
@@ -1100,52 +1097,71 @@ was matched (see :class:`pyramid.interfaces.IRoute` for the API of such a route
...
 .. code-block:: python
     :linenos:
 
-    def any_of(segment_name, *allowed):
-        def predicate(info, request):
-            if info['match'][segment_name] in allowed:
-                return True
-        return predicate
+    class AnyOfPredicate:
+        def __init__(self, val):
```
Comment (mmerickel) [inline docs/narr/urldispatch.rst]: predicates take two args, `(val, info)` where the second arg is an `IPredicateInfo` object

File: docs/narr/urldispatch.rst (hunk 1 above)
Comment (mmerickel) [reply docs/narr/urldispatch.rst]: It looks good aside from my comment above about the init signature.

File: src/pyramid/config/routes.py (hunk 18 above)
Comment (mmerickel) [reply src/pyramid/config/routes.py]: Worse case scenario I'd just use backticks to escape it.

```
``**predicates``
```

Comment (stevepiercy) [reply src/pyramid/config/routes.py]: That was one combination I did not try. Double-escaping works! `\\*\\*predicates`.

**Comment by merwok** [comment]
I will wrap this up today.

File: docs/quick_tour.rst [hunk 20]
```diff
@@ -478,12 +478,12 @@ more to offer:
...
 
 .. seealso:: See also:
     :ref:`Quick Tutorial View Classes <qtut_view_classes>`, :ref:`Quick
-    Tutorial More View Classes <qtut_more_view_classes>`, and
-    :ref:`class_as_view`.
+    Tutorial More View Classes <qtut_more_view_classes>`, :ref:`class_as_view`,
+    :ref:`view_and_route_predicates`.
```
[the remark this answers is not in the document]
Comment (merwok) [reply docs/quick_tour.rst]: Not changing it: the title in the document is `09: Organizing Views With View Classes`, which reads a bit awkward.  What do you think of:

```diff
 .. seealso:: See also:
-    :ref:`Quick Tutorial View Classes <qtut_view_classes>`, :ref:`Quick
-    Tutorial More View Classes <qtut_more_view_classes>`, :ref:`class_as_view`,
-    :ref:`view_and_route_predicates`.
+    :ref:`View Classes <qtut_view_classes>` and :ref:`More View Classes <qtut_more_view_classes>`
+    in the Quick Tutorial, :ref:`class_as_view`, and :ref:`view_and_route_predicates`.
```

File: docs/narr/urldispatch.rst (hunk 19 above)
Comment (merwok) [reply docs/narr/urldispatch.rst]: Ah!  I translated the old example, mapping the wrapper function to predicate factory (`__init__`) and the actual predicate function to `__call__`, but I see now that the `ContentTypePredicate` example in `hooks.rst` does have the signature that you mention.
Also the `info` param passed to the predicate factory implements IPredicateInfo, but `info` passed to route predicate instances is a dict with route match details.  It’s clear now!

File: HISTORY.rst [hunk 21]
```diff
@@ -2530,7 +2530,7 @@ Features
      @view_config(abc=1)
 
   Similar features exist for ``add_route``, and ``add_subscriber``.  See
-  "Adding A Third Party View, Route, or Subscriber Predicate" in the Hooks
+  "Adding A Custom View, Route, or Subscriber Predicate" in the Hooks
```
[the remark this answers is not in the document]
Comment (merwok) [reply HISTORY.rst]: I guess the reason is that HISTORY was included in docs but also in package long description (see https://github.com/Pylons/pyramid/pull/3563#discussion_r366588973).  I had changed my local copy to use `ref`, but it would be the only line in the file to do so, and render it dependent on Sphinx rather than pure reST, so I am reverting now (to the version with `Custom`, so that if anyone reads line 2533 they’ll have a chance at a successful docs search :slightly_smiling_face:)

File: docs/narr/urldispatch.rst (hunk 19 above)
Comment (merwok) [reply docs/narr/urldispatch.rst]: I feel that there should be at least a very short mention of this `info` param in the doc below, even though it’s already explained in the full docs in `hooks.rst`.

File: HISTORY.rst (hunk 21 above)
[the remark this answers is not in the document]
Comment (mmerickel) [reply HISTORY.rst]: Yes the CHANGES is copied to HISTORY upon release of a major version. They should both minimize any linking to the rest of the document. The whatsnew docs will link back to them though to display the raw changelog.

File: src/pyramid/config/routes.py [hunk 22]
```diff
@@ -40,12 +40,9 @@ def add_route(
...
     ):
-        r""" Add a :term:`route configuration` to the current
-        configuration state, as well as possibly a :term:`view
-        configuration` to be used to specify a :term:`view callable`
-        that will be invoked when this route matches.  The arguments
-        to this method are divided into *predicate*, *non-predicate*,
-        and *view-related* types.  :term:`Route predicate` arguments
+        r""" Add a :term:`route configuration` to the current configuration state.
+        The arguments to this method are divided into *predicate* and
+        *non-predicate* types.  :term:`Route predicate` arguments
```
[the remark this answers is not in the document]
Comment (merwok) [reply src/pyramid/config/routes.py]: I have found that some whitespace things are not changed by black but caught by flake8.  Not sure about docstrings; I am used to PEP 257 conventions so I format to 72 columns before I run linters.

File: docs/narr/urldispatch.rst (hunk 19 above)
Comment (merwok) [reply docs/narr/urldispatch.rst]: In tests, predicate classes are defined with ``__init__(self, val, config)``, maybe to avoid the info/info confusion.  I could follow this convention, even though the interface is IPredicateInfo.

**Comment by merwok** [comment]
- [ ]  one small wording issue open
- [ ]  question about info/config param
- [ ]  should probably test the example code just to be sure

File: docs/narr/urldispatch.rst (hunk 19 above)
Comment (mmerickel) [reply docs/narr/urldispatch.rst]: config is legacy because it used to be the actual configurator, but I don't really care what you call it

File: HISTORY.rst (hunk 21 above)
[the remark this answers is not in the document]
Comment (stevepiercy) [reply HISTORY.rst]: tox has a lint test for this as well: https://github.com/Pylons/pyramid/blob/master/tox.ini#L24

File: src/pyramid/config/routes.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (stevepiercy) [reply src/pyramid/config/routes.py]: https://pylonsproject.org/community-coding-style-standards.html

I need to update some of the Coding Style standards since we started using black ourselves. I haven't audited that. Another todo for that project. Pyramid's linters are the current standard.

File: docs/quick_tour.rst (hunk 20 above)
[the remark this answers is not in the document]
Comment (stevepiercy) [reply docs/quick_tour.rst]: I like that.

**Comment by merwok** [comment]
I’ll run a single-file test!

**Comment by merwok** [comment]
The code seems to work well: https://gist.github.com/merwok/c17b11dfa4e462efc6ea0d8ea9585c0c

Running a linter found a warning, now fixed.
